metricbeat/module/redis/metricset.go

```diff
@@ -50,9 +50,16 @@ func NewMetricSet(base mb.BaseMetricSet) (*MetricSet, error) {
 		return nil, errors.Wrap(err, "failed to read configuration")
 	}
 
+	var password string
+	if base.HostData().Password != "" {
+		password = base.HostData().Password
+	} else {
+		password = config.Password
```

Comment:
If you look at `URLHostParserBuilder` you'll see that it takes into account the value of the `password` config option when the URL doesn't contain a password. That should make this code block unnecessary. Likewise `Password` can be removed from the struct declaration above (like 42).

Then down below you can pass `base.HostData().Password` into `CreatePool`. Try that and make sure it's working like I think it supposed to 😉 .

Reply:
I don't see the password being populated as you describe:
a) With this config:
```yaml
- module: redis
  period: 10s
  hosts: ["redis://127.0.0.1:6379"]
  password: testpassword
```
These are the values:
- `base.HostData().Password` is empty.
- `config.Password` is `testpassword`

b) With this config:
```yaml
- module: redis
  period: 10s
  hosts: ["redis://:newpassword@127.0.0.1:6379"]
  password: testpassword
```
These are the values:
- `base.HostData().Password` is `newpassword` (in my changes we prioritize this one, the other hosts would use `config.Password`).
- `config.Password` is `testpassword`

c) With this config:
```yaml
- module: redis
  period: 10s
  hosts: ["127.0.0.1:6379"]
  password: testpassword
```
These are the values:
- `base.HostData().Password` is empty.
- `config.Password` is `testpassword`.

Reply:
I found the problem have provided a patch that you can incorporate. The issue was that the URL parser was not setting the password when there was no username. I also added a test case to verify the behavior or the redis metricset.

[url-parse-password.patch.txt](https://github.com/elastic/beats/files/2928217/url-parse-password.patch.txt)

Reply:
Great, thanks!